### DISCUSSION TITLE
How can I test locally using `sam local start-lambda` with a Rust lambda?

### DISCUSSION THREAD

**Discussion by shirakaba** [aws/aws-lambda-rust-runtime#590] [Q&A] [answered]
I've set up a lambda with CDK by referring to the repository structure used by [rnag/rust.aws-cdk-lambda/cdk-examples/rust-standalone](https://github.com/rnag/rust.aws-cdk-lambda/tree/main/cdk-examples/rust-standalone), whilst matching this repo's up-to-date dependencies.

I want to do local testing (run this lambda as a server on localhost and make HTTP GET requests to it), so I've tried running this:

```sh
$ sam local start-lambda --port 3000 --template ./cdk.out/RustStandaloneStack.template.json   
Starting the Local Lambda Service. You can now invoke your Lambda Functions defined in your template through the endpoint.
2023-01-23 11:08:56  * Running on http://127.0.0.1:3000/ (Press CTRL+C to quit)
```

The server successfully starts up, but I can't figure out what sort of request to send it. It just responds with a 404 no matter what path or headers I specify to it with my GET requests.

By comparison, one-off invocations work just fine:

```sh
# ./tests/sample.json is an event matching the same format of the one seen in this issue comment:
# https://github.com/awslabs/aws-lambda-rust-runtime/issues/365#issue-1054795934
$ sam local invoke MyRustLambda --event ./tests/sample.json --template ./cdk.out/RustStandaloneStack.template.json
```

Requesting using the exact same path and headers specified in my `./tests/sample.json` file, I am met with a 404 response.

This feels like an underdocumented area in the SAM/CDK/Lambda docs so any pointers would be greatly appreciated.

**Comment by calavera** [comment]
I'm not familiar with `sam local`, but these links might help you test locally:

https://www.cargo-lambda.info/commands/watch.html
https://www.cargo-lambda.info/commands/invoke.html
https://www.cargo-lambda.info/guide/screencasts.html#cargo-lambda-watch

**Comment by bnusunny** [comment]
`sam local start-lambda` starts up a local Lambda Service endpoint. You can find out the request format from [Lambda API doc](https://docs.aws.amazon.com/lambda/latest/dg/API_Invoke.html). You can send POST request to this endpoint via curl or AWS SDK.

AWS CLI is a simpler way to invoke lambda via this endpoint. More details [here](https://awscli.amazonaws.com/v2/documentation/api/2.3.2/reference/lambda/invoke.html). 

```bash
aws lambda invoke --function-name "HelloWorldFunction" --endpoint-url "http://127.0.0.1:3001" --no-verify-ssl out.txt
```

**Comment by shirakaba** [comment, marked as answer]
Thank you both very much for the timely response that let me get straight back to work! I'll check through @calavera's resources as well in a moment, but @bnusunny's Lambda API docs look like exactly the piece of the puzzle I've been missing.

# Recap

To recap, I'm trying to translate the following `sam local invoke` to a `sam local start-lambda` + HTTP request (for testing a web app):

```sh
sam local invoke MyRustLambda --event ./tests/sample.json --template ./cdk.out/RustStandaloneStack.template.json
```

... where `tests/sample.json` could be as complex [as this example](https://github.com/awslabs/aws-lambda-rust-runtime/issues/365#issue-1054795934) or as simple as just:

```json
{
  "requestContext": {
    "http": {
      "method": "GET"
    },
    "timeEpoch": 1674445263
  }
}
```

# Step one: `sam local start-lambda`

We begin by starting up the lambda as a HTTP server:

```sh
sam local start-lambda --port 3000 --template ./cdk.out/RustStandaloneStack.template.json
```

# Step two: the HTTP request

## Background

Looking at the [Lambda API doc](https://docs.aws.amazon.com/lambda/latest/dg/API_Invoke.html), it says the POST request should be in the following format:

```
POST /2015-03-31/functions/FunctionName/invocations?Qualifier=Qualifier HTTP/1.1
X-Amz-Invocation-Type: InvocationType
X-Amz-Log-Type: LogType
X-Amz-Client-Context: ClientContext
```

## Conclusion

So in my case, as I can leave `X-Amz-Client-Context` blank (given that my Lambda doesn't make use of it), my HTTP request would become:

```sh
curl -X POST --data @./tests/sample.json \
  --header "X-Amz-Invocation-Type: RequestResponse" \
  --header "X-Amz-Log-Type: None" \
  http://localhost:3000/2015-03-31/functions/MyRustLambda/invocations
```

And as the headers are inconsequential (the first header asks for the default value and I don't mind whether I get logs or not), I can simplify that further to just:

```sh
curl -X POST --data @./tests/sample.json http://localhost:3000/2015-03-31/functions/MyRustLambda/invocations
```
